File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/vehicle/AirFrictionForceCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.vehicle;

import com.github.aadvorak.artilleryonline.battle.calculations.ForceAtPoint;
import com.github.aadvorak.artilleryonline.battle.calculations.ForceCalculator;
import com.github.aadvorak.artilleryonline.battle.calculations.VehicleCalculations;
import com.github.aadvorak.artilleryonline.battle.common.Force;
import com.github.aadvorak.artilleryonline.battle.config.VehicleConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;
import com.github.aadvorak.artilleryonline.battle.precalc.VehiclePreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.VehicleSpecs;
import com.github.aadvorak.artilleryonline.battle.state.VehicleState;

import java.util.List;

public class AirFrictionForceCalculator implements ForceCalculator<
        VehicleSpecs,
        VehiclePreCalc,
        VehicleConfig,
        VehicleState,
        VehicleModel,
        VehicleCalculations> {

    @Override
    public List<ForceAtPoint> calculate(VehicleCalculations calculations, BattleModel battleModel) {
        var frictionCoefficient = battleModel.getRoom().getSpecs().getAirFrictionCoefficient();
        var velocity = calculations.getVelocity();
        // todo remove mass with size coefficient
        var mass = calculations.getMass();
        var frictionForce = new Force()
                .setX( - velocity.getX() * Math.abs(velocity.getX()) * frictionCoefficient * mass)
                .setY( - velocity.getY() * Math.abs(velocity.getY()) * frictionCoefficient * mass);
        return List.of(ForceAtPoint.atCOM(frictionForce));
    }
}

File: frontend/playground/battle/calculator/vehicle/air-friction-force-calculator.ts
import type {ForceCalculator} from "~/playground/battle/calculator/force-calculator";
import type {VehicleCalculations} from "~/playground/data/calculations";
import type {BattleModel} from "~/playground/data/model";
import {BodyForce} from "~/playground/battle/calculator/body-force";

export class AirFrictionForceCalculator implements ForceCalculator {
  private static readonly FORCE_DESCRIPTION = 'Air Friction'

  calculate(calculations: VehicleCalculations, battleModel: BattleModel): BodyForce[] {
    const frictionCoefficient = battleModel.room.specs.airFrictionCoefficient
    const velocity = calculations.model.state.velocity
    // todo remove mass with size coefficient
    const mass = calculations.model.preCalc.mass

    const frictionForce = {
      x: -velocity.x * Math.abs(velocity.x) * frictionCoefficient * mass,
      y: -velocity.y * Math.abs(velocity.y) * frictionCoefficient * mass
    }

    return [BodyForce.atCOM(frictionForce, AirFrictionForceCalculator.FORCE_DESCRIPTION)]
  }
}
